lib/DocumentManagerContext.php

```diff
@@ -0,0 +1,199 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\DocumentManager;
+
+use PHPCR\SessionInterface;
+use ProxyManager\Factory\LazyLoadingGhostFactory;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * The document manager context is a container for all the dependencies of a
+ * given document manager.
+ */
+class DocumentManagerContext
+{
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var DocumentInspectorFactory
+     */
+    private $inspectorFactory;
+
+    /**
+     * @var NodeManager
+     */
+    private $nodeManager;
+
+    /**
+     * @var DocumentRegistry
+     */
+    private $registry;
+
+    /**
+     * @var ProxyFactory
+     */
+    private $proxyFactory;
+
+    /**
+     * @var MetadataFactoryInterface
+     */
+    private $metadataFactory;
+
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $manager;
+
+    /**
+     * @var string
+     */
+    private $name;
+
+    public function __construct(
+        $name,
+        SessionInterface $session,
+        EventDispatcherInterface $eventDispatcher,
+        MetadataFactoryInterface $metadataFactory,
+        LazyLoadingGhostFactory $lazyProxyFactory,
+        DocumentInspectorFactoryInterface $inspectorFactory,
+        $defaultLocale
+    ) {
+        $this->name = $name;
+
+        // the event dispatcher should be unique to this document manager context
+        // instance.
+        $this->eventDispatcher = $eventDispatcher;
+
+        // the inspector factory provides a way for users to instantiate
+        // their own document inspectors.
+        $this->inspectorFactory = $inspectorFactory;
+
+        // the metadata factory is currently intended to be the same for all
+        // document mnaager contexts instances.
+        $this->metadataFactory = $metadataFactory;
+
+        // the PHPCR session SHOULD be unique to all (end user) document
+        // manager contexts.
+        $this->session = $session;
+
+        // instantiate other objects scoped to this document manager.
+        $this->nodeManager = new NodeManager($session);
+        $this->proxyFactory = new ProxyFactory($this, $lazyProxyFactory, $metadataFactory);
+        $this->registry = new DocumentRegistry($defaultLocale);
+        $this->manager = new DocumentManager($this);
```

Comment:
For example, as you suggest, we would here inject the PHPCR session and the NodeManager (which contains the PHPCR session) `__construct($phpcrSession, $nodeManager, ..)`. 

It would make no sense for the NodeManager to have a Session which is different from the PhpcrSession we provide to the context.

Otherwise: 

1. it provides a clear API for the dependencies of the context. To create a document manager instance, you just need to instantiate the Context, you do not need to instantiate several interdependent object instances - and potentially fuck it up.
2. The DocumentManager requires an instance of the `Context`, which requires an instance of the `DocumentManager`. Making the `Context` a factory for the `DocumentManager` is better than playing some strange `chicken and egg` game (the reason for the previous `Context::attachDocumentManager` method in previous PR version).

I have considered, and tried, instantaiting everything, but it really doesn't make sense to me for the above reasons.